### PR TITLE
Pin gdal 2.1.0

### PR DESCRIPTION
**Pull request by ocefpaf**
Prevents from wrongly downloading `gdal` from the default channel.

### CODE REVIEW

**Comment by phobson** [comment]
Is this related to problems I'm seeing with fiona in general? e.g., https://travis-ci.org/Geosyntec/pygridtools/jobs/141003683#L541

**Comment by ocefpaf** [comment]
@gillins I am seeing the error bellow when using the latest `gdal 2` build (`gdal 2.1.0-np110py27_6`)

Any advice?

```shell
===== testing package: fiona-1.7.0-np110py27_1 =====
import: 'fiona'
Traceback (most recent call last):
  File "/Users/travis/miniconda3/conda-bld/test-tmp_dir/run_test.py", line 25, in <module>
    import fiona
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/__init__.py", line 69, in <module>
    from fiona.collection import Collection, BytesCollection, vsi_path
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/collection.py", line 7, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so, 2): Library not loaded: /Users/travis/miniconda3/conda-bld/work/gdal-2.1.0/libgdal.dylib
  Referenced from: /Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so
  Reason: image not found
TESTS FAILED: fiona-1.7.0-np110py27_1
```

**Comment by gillins** [comment]
Same thing I'm seeing here: https://github.com/conda-forge/gdal-feedstock/issues/81

I have no ideas I'm afraid...

**Comment by ocefpaf** [comment]
> Same thing I'm seeing here: https://github.com/conda-forge/gdal-feedstock/issues/81
> I have no ideas I'm afraid...

I lost track of the changes from build 5 to build 6 on `gdal`. Would you be OK removing build 6 from the channel for now? Or is that even more harmful?

**Comment by gillins** [comment]
Fine with me. Just seems to be OSX.

**Comment by ocefpaf** [comment]
> Fine with me. Just seems to be OSX.

Thanks! I will investigate this more once I get some free time. Hopefully `build 5` will be OK for most of our users.

**Comment by ocefpaf** [comment]
> OS X has been broken for a while now with the whole `libc++`/`libstdc++` thing anyway.

Yeah. I lost most of the discussion on that front. Not sure if the `toolchain` is the answer though. Do you know any build number where that was not an issue? Or has the `libc++`/`libstdc++` issue always there from the beginning?

**Comment by gillins** [comment]
I'm afraid not - maybe before numpy+openblas? I think the `libc++`/`libstdc++` issue has always been there, just hasn't tripped us up before for some reason. There seems to be a lot of OSX specific knowledge needed for this that I don't have... Are there many people using Conda on OSX do you think? We need someone experienced with OSX to come along and sort all the problems out.

**Comment by phobson** [comment]
It's not just OS X though. I'm seeing this on Linux through TravisCI and a Linux Mint VM (via VirtualBox)

**Comment by ocefpaf** [comment]
> It's not just OS X though. I'm seeing this on Linux through TravisCI and a Linux Mint VM (via VirtualBox)

The `libc++/libstdc++` issue?

> I'm afraid not - maybe before numpy+openblas?

Yeah. The numpy+openblas played a number on us...

**Comment by ocefpaf** [comment]
I feel like playing whack-a-mole here. Now it is `kealib`. The only difference I can spot is the `toolchain`. That is the second package it breaks :confused: 

I will experiment removing that and re-adding if it turns out not the real issue.

```shell
import: 'fiona'
Traceback (most recent call last):
  File "/Users/travis/miniconda3/conda-bld/test-tmp_dir/run_test.py", line 25, in <module>
    import fiona
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/__init__.py", line 69, in <module>
    from fiona.collection import Collection, BytesCollection, vsi_path
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/collection.py", line 7, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so, 2): Symbol not found: __ZN6kealib10KEAImageIO10isKEAImageESs
  Referenced from: /Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/../../..//libgdal.20.dylib
  Expected in: /Users/travis/miniconda3/envs/_test/lib//libkea.1.4.6.dylib
 in /Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/../../..//libgdal.20.dylib
TESTS FAILED: fiona-1.7.0-np110py27_1
```

**Comment by ocefpaf** [comment]
There is no end to the problem. Now it is `hdf5`. I cannot keep deleting builds that use the `toolchain` in order to restore the healthy builds. We need a better solution for this!

```shell
    import fiona
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/__init__.py", line 69, in <module>
    from fiona.collection import Collection, BytesCollection, vsi_path
  File "/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/collection.py", line 7, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/travis/miniconda3/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so, 2): Symbol not found: __ZN2H56H5FileC1ERKSsjRKNS_17FileCreatPropListERKNS_15FileAccPropListE
  Referenced from: /Users/travis/miniconda3/envs/_test/lib//libkea.1.4.6.dylib
  Expected in: /Users/travis/miniconda3/envs/_test/lib//libhdf5_cpp.12.dylib
 in /Users/travis/miniconda3/envs/_test/lib//libkea.1.4.6.dylib
TESTS FAILED: fiona-1.7.0-np110py27_1
```

**Comment by ocefpaf** [comment]
> This is getting ridiculous. We are trying to find solutions to problems, which weren't there before the `toolchain` was introduced and probably won't be there without it. If this is getting into packages like hdf5, which is widely used as dependency, this will end in total frustration.

Ping @pelson.

**Comment by kmuehlbauer** [comment]
Having said that, I'm not against a `toolchain`. But it should just work and not introduce *hard to debug* errors.

**Comment by pelson** [comment]
So IMO I haven't seen the cause of the problem, so all I know at this moment is the effects of the problem (it is broken for the user == not good), and one solution (to remove all of the broken builds). 

@ocefpaf - would it be worth doing a quick hangout to talk about the potential underlying issue? (likely either bad pinning, or a binary incompatibility introduced by changing a build flag. Whilst I don't know what that could be, one example of something similar would be if we changed the wide vs narrow flag when compiling python).

**Comment by ocefpaf** [comment]
> would it be worth doing a quick hangout to talk about the potential underlying issue? 

Sure. Ideally we should talk with @gillins, but I don't think that will be possible today. Ping me whenever you are ready.

**Comment by pelson** [comment]
Looks like fiona itself is looking for a different version of gdal:

```
$ ldd lib/python2.7/site-packages/fiona/ogrext.so 
	linux-vdso.so.1 =>  (0x00007ffd8c1e5000)
	** libgdal.so.20 => not found **
	libpython2.7.so.1.0 => lib/python2.7/site-packages/fiona/../../../libpython2.7.so.1.0 (0x00007f72e1a94000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f72e1877000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f72e14e3000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f72e12de000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007f72e10db000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f72e0e57000)
	/lib64/ld-linux-x86-64.so.2 (0x0000003670800000)
```

But actually there is no `libgdal.so.20`:

```
$ ls -l lib/libgdal.so*  
lib/libgdal.so -> libgdal.so.2.1.0
lib/libgdal.so.2 -> libgdal.so.2.1.0
lib/libgdal.so.2.1.0
```

This suggests the fiona pinning of ``gdal >=2.*`` is not strict enough, and should perhaps be `gdal 2.1.*` (note: I have not looked at the naming convention for the gdal `.so`, so may need to be slightly altered).

**Comment by ocefpaf** [comment]
> This suggests the fiona pinning of ``gdal >=2.*`` is not strict enough, and should perhaps be `gdal 2.1.*` (note: I have not looked at the naming convention for the gdal `.so`, so may need to be slightly altered).

That is the purpose of this PR. Or am I missing something?

**Comment by ocefpaf** [comment]
Also, if you look at https://github.com/conda-forge/fiona-feedstock/pull/17#issuecomment-229697339 you'll see that the linking issue now are with `libhdf5_cpp.12.dylib` and `libkea.1.4.6.dylib`. Both issues appeared after we compiled `kealib` and `hdf5` with the `toolchain`.

**Comment by pelson** [comment]
> That is the purpose of this PR. Or am I missing something?

I think the point is that *I* was missing something - I believe there is now good evidence (i.e. documented in this issue for the future) for pinning GDAL down to minor version (in `major.minor.patch` terms).

The dylib problem looks like a C++ name mangling problem which @jakirkham may be in a position to have more insight on.

**Comment by pelson** [comment]
FWIW, I'm able to reproduce this on my machine:

```
===== testing package: fiona-1.7.0-np111py27_1 =====
import: u'fiona'
Traceback (most recent call last):
  File "/Users/pelson/miniconda/conda-bld/test-tmp_dir/run_test.py", line 26, in <module>
    import fiona
  File "/Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/__init__.py", line 69, in <module>
    from fiona.collection import Collection, BytesCollection, vsi_path
  File "/Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/collection.py", line 7, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/ogrext.so, 2): Symbol not found: __ZN6kealib10KEAImageIO10isKEAImageESs
  Referenced from: /Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/../../..//libgdal.20.dylib
  Expected in: /Users/pelson/miniconda/envs/_test/lib//libkea.1.4.6.dylib
 in /Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/../../..//libgdal.20.dylib
TESTS FAILED: fiona-1.7.0-np111py27_1
```

```
otool -L /Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/*.so /Users/pelson/miniconda/envs/_test/lib/*.dylib | grep stdc -B 10

/Users/pelson/miniconda/envs/_test/lib/python2.7/site-packages/fiona/_transform.so:
	@rpath/libgdal.20.dylib (compatibility version 22.0.0, current version 22.0.0)
	@rpath/./libstdc++.6.dylib (compatibility version 7.0.0, current version 7.19.0)

/Users/pelson/miniconda/envs/_test/lib/libgdal.20.dylib:
	@rpath/libgdal.20.dylib (compatibility version 22.0.0, current version 22.0.0)
	@rpath/libjson-c.2.dylib (compatibility version 3.0.0, current version 3.1.0)
	@rpath/libexpat.1.dylib (compatibility version 8.0.0, current version 8.0.0)
	@rpath/./libxerces-c-3.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libopenjp2.7.dylib (compatibility version 7.0.0, current version 2.1.0)
	@rpath/libnetcdf.11.dylib (compatibility version 11.0.0, current version 11.0.0)
	@rpath/libmfhdf.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libdf.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libgif.7.dylib (compatibility version 8.0.0, current version 8.0.0)
	@rpath/libtiff.5.dylib (compatibility version 8.0.0, current version 8.4.0)
	@rpath/libjpeg.9.dylib (compatibility version 12.0.0, current version 12.0.0)
	@rpath/libpng16.16.dylib (compatibility version 38.0.0, current version 38.0.0)
	@rpath/libpq.5.dylib (compatibility version 5.0.0, current version 5.4.0)
	@rpath/libspatialite.7.dylib (compatibility version 9.0.0, current version 9.0.0)
	@rpath/libfreexl.1.dylib (compatibility version 3.0.0, current version 3.0.0)
	@rpath/libcharset.1.dylib (compatibility version 2.0.0, current version 2.0.0)
	@rpath/libproj.9.dylib (compatibility version 11.0.0, current version 11.0.0)
	@rpath/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	@rpath/libgeos_c.1.dylib (compatibility version 10.0.0, current version 10.2.0)
	@rpath/libgeos-3.4.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libcurl.4.dylib (compatibility version 9.0.0, current version 9.0.0)
	@rpath/libssl.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libxml2.2.dylib (compatibility version 12.0.0, current version 12.3.0)
	@rpath/libicui18n.56.dylib (compatibility version 56.0.0, current version 56.1.0)
	@rpath/libicuuc.56.dylib (compatibility version 56.0.0, current version 56.1.0)
	@rpath/libicudata.56.dylib (compatibility version 56.0.0, current version 56.1.0)
	@rpath/liblzma.5.dylib (compatibility version 6.0.0, current version 6.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
	@rpath/libiconv.2.dylib (compatibility version 8.0.0, current version 8.1.0)
	@rpath/libkea.1.4.6.dylib (compatibility version 1.4.6, current version 1.4.6)
	@rpath/libhdf5_hl.10.dylib (compatibility version 12.0.0, current version 12.0.0)
	@rpath/libhdf5_cpp.12.dylib (compatibility version 13.0.0, current version 13.0.0)
	@rpath/libhdf5.10.dylib (compatibility version 13.0.0, current version 13.0.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.8)
	@rpath/./libstdc++.6.dylib (compatibility version 7.0.0, current version 7.19.0)

--

/Users/pelson/miniconda/envs/_test/lib/libgeos.dylib:
	libgeos-3.4.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 60.0.0)
--
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
	/usr/lib/libgcc_s.1.dylib (compatibility version 1.0.0, current version 2577.0.0)


/Users/pelson/miniconda/envs/_test/lib/libgeos_c.dylib:
	libgeos_c.1.dylib (compatibility version 10.0.0, current version 10.2.0)
	@loader_path/./libgeos-3.4.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 60.0.0)

/Users/pelson/miniconda/envs/_test/lib/libstdc++.dylib:
	@rpath/./libstdc++.6.dylib (compatibility version 7.0.0, current version 7.19.0)
--
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)

/Users/pelson/miniconda/envs/_test/lib/libxerces-c-3.1.dylib:
	@rpath/./libxerces-c-3.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 53.0.0)
	/usr/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 52.0.0)
--
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 635.21.0)


/Users/pelson/miniconda/envs/_test/lib/libxerces-c.dylib:
	@rpath/./libxerces-c-3.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 53.0.0)
	/usr/lib/libstdc++.6.dylib (compatibility version 7.0.0, current version 52.0.0)
```


```
conda list -n _test
# packages in environment at /Users/pelson/miniconda/envs/_test:
#
click                     6.6                      py27_0    defaults
cligj                     0.4.0                    py27_0    defaults
ca-certificates           2016.2.28                     1    conda-forge
click-plugins             1.0.3                    py27_0    conda-forge
curl                      7.49.1                        0    conda-forge
freexl                    1.0.2                         1    conda-forge
gdal                      2.1.0               np111py27_5    conda-forge
giflib                    5.1.2                         1    conda-forge
hdf4                      4.2.11                        4    conda-forge
hdf5                      1.8.17                        2    conda-forge
icu                       56.1                          4    conda-forge
jpeg                      9b                            0    conda-forge
json-c                    0.12                          0    conda-forge
kealib                    1.4.6                         2    conda-forge
libiconv                  1.14                          2    conda-forge
libnetcdf                 4.4.0                         1    conda-forge
libpng                    1.6.23                        0    conda-forge
libspatialite             4.3.0a                        8    conda-forge
libtiff                   4.0.6                         5    conda-forge
libxml2                   2.9.3                         8    conda-forge
munch                     2.0.4                    py27_0    conda-forge
ncurses                   5.9                           7    conda-forge
nose                      1.3.7                    py27_0    conda-forge
openblas                  0.2.18                        2    conda-forge
openjpeg                  2.1.0                         5    conda-forge
openssl                   1.0.2h                        0    conda-forge
proj.4                    4.9.2                         0    conda-forge
readline                  6.2                           0    conda-forge
six                       1.10.0                   py27_0    conda-forge
sqlite                    3.13.0                        1    conda-forge
xz                        5.0.5                         1    conda-forge
zlib                      1.2.8                         3    conda-forge
libgcc                    4.8.5                         1    defaults
blas                      1.1                    openblas    local
expat                     2.1.0                         3    local
fiona                     1.7.0               np111py27_1    local
geos                      3.4.2                         0    local
krb5                      1.14.2                        0    local
numpy                     1.11.0          py27_blas_openblas_100  [blas_openblas]  local
setuptools                18.4                    py27_21    local
postgresql                9.1.4                         1    defaults
python                    2.7.12                        0    defaults
tk                        8.5.18                        0    defaults
xerces-c                  3.1.2                         0    defaults
```

@jakirkham - a big problem with regards to https://github.com/conda-forge/conda-forge.github.io/issues/183 is that we are actually installing ``libstdc++.dylib`` as part of an installation of openblas, and so we are polluting the environment with it before we even start trying to isolate it from individual distributions. With this in mind, shall we move numpy into a `dev` label (i.e. on `conda-forge/label/dev`) so that we avoid polluting the build and test environments in this way? With that done, we can start to whittle down the packages which show-up above.

In summary, I think we will need to re-build (and potentially modify the build process) of:

 * [ ] - fiona
 * [ ] - geos
 * [ ] - gdal

Additionally, presumably we are going to see problems from xerces-c (gdal dependency) as it links libstdc++ but we do not build it on conda-forge.

Even if we aren't able to solve the problem immediately, I'd love to lay out a plan to solve this issue. If we aren't able to see a path through this in the short term, we may be better off reverting back to the old `clang` `gcc` front-end for the time being.

**Comment by jakirkham** [comment]
On the `libstdc++` being pulled in via `libgcc` on Mac, this [comment]( https://github.com/conda-forge/gdal-feedstock/pull/85#issuecomment-230006423 ) is relevant. Basically the problem isn't with `libstdc++` we use. It is that we are using it at all. We should be using `libc++`. As soon as we start doing that, things are going to get better for us.

Though we should definitely resume `libgfortran` for Mac ASAP and `libgomp` for both as lobbing all the compiler libraries in one `libgcc` package is getting silly. Note this has long range benefit no matter how we build stuff.

So, @gillins already did `geos` see this PR ( https://github.com/conda-forge/geos-feedstock/pull/16 ). It looks like @kmuehlbauer is having some success with `gdal` as seen in this PR ( https://github.com/conda-forge/gdal-feedstock/pull/85 ). We should double check it to make sure all the builds are clean. That leaves `fiona` and `rasterio` that still need addressing.

**Comment by TimoRoth** [comment]
When I faced the same problem in our project repository, i fixed(workarounded?) it by adding "--channel conda-forge --channel defaults --override-channels" to all calls to conda build.
It makes it prefer the conda-forge repository, only falling back to defaults if something is missing on conda-forge.
The order is important here, conda-forge has to come first, otherwise it will prefer default, which is not desirable and breaks a lot of stuff.

**Comment by ocefpaf** [comment]
> When I faced the same problem in our project repository, i fixed(workarounded?) it by adding "--channel conda-forge --channel defaults --override-channels" to all calls to conda build.
It makes it prefer the conda-forge repository, only falling back to defaults if something is missing on conda-forge.
The order is important here, conda-forge has to come first, otherwise it will prefer default, which is not desirable and breaks a lot of stuff.

@TimoRoth we had a collection of issues here that are (hopefully) solved. However, updating `conda` and adding `conda-forge` on top of `defaults` in your `.condarc` file is a good idea to take advantage of the channel preference feature.
